README.md
# Proxywrap ![test-badge](http://strider.findhit.com/findhit/findhit-proxywrap/badge)

This module is a fork of original [proxywrap](https://github.com/daguej/node-proxywrap) by [Josh Dague](https://github.com/daguej). Unfortunately, the project doesn't have recent changes. As so, we decided to contribute to it by forking it and make it better. Do you have any idea to improve this? Feel free to open an **Issue** or **Pull Request**.

This module wraps node's various `Server` interfaces so that they are compatible with the [PROXY protocol](http://haproxy.1wt.eu/download/1.5/doc/proxy-protocol.txt).  It automatically parses the PROXY headers and resets `socket.remoteAddress` and `socket.remotePort` so that they have the correct values.

```

    npm install findhit-proxywrap

```

This module is especially useful if you need to get the client IP address when you're behind an AWS ELB in TCP mode.

In HTTP or HTTPS mode (aka SSL termination at ELB), the ELB inserts `X-Forwarded-For` headers for you.  However, in TCP mode, the ELB can't understand the underlying protocol, so you lose the client's IP address.  With the PROXY protocol and this module, you're able to retain the client IP address with any protocol.

In order for this module to work, you must [enable the PROXY protocol on your ELB](http://docs.aws.amazon.com/ElasticLoadBalancing/latest/DeveloperGuide/enable-proxy-protocol.html) (or whatever proxy your app is behind).

Usage
-----

proxywrap is a drop-in replacement.  Here's a simple Express app:

```

    var http = require('http')
        , proxiedHttp = require('findhit-proxywrap').proxy( http )
        , express = require('express')
        , app = express()
        , srv = proxiedHttp.createServer(app); // instead of http.createServer(app)

    app.get('/', function(req, res) {
        res.send('IP = ' + req.connection.remoteAddress + ':' + req.connection.remotePort);
    });

    srv.listen(80);

```

The magic happens in the `proxywrap.proxy()` call.  It wraps the module's `Server` constructor and handles a bunch of messy details for you.

You can do the same with `net` (raw TCP streams), `https`, and `spdy`.  It will probably work with other modules that follow the same pattern, but none have been tested.

*Note*: If you're wrapping [node-spdy](https://github.com/indutny/node-spdy), its exports are a little strange:

    var proxiedSpdy = require('proxywrap').proxy(require('spdy').server);

This also adds to all your sockets the properties:
* `socket.clientAddress` - The IP Address that connected to your PROXY.
* `socket.clientPort` - The Port used by who connected to your PROXY.
* `socket.proxyAddress` - The IP Address exposed on Client <-> Proxy side.
* `socket.proxyPort` - The Port exposed on Client <-> Proxy side. Usefull for detecting SSL on AWS ELB.
* `socket.remoteAddress` [optional] - Same as `socket.clientAddress`, used for compability proposes.
* `socket.remotePort` [optional] - Same as `socket.clientPort`, used for compability proposes.

**Warning:** By default, *all* traffic to your proxied server MUST use the PROXY protocol.  If the first five bytes received aren't `PROXY`, the connection will be dropped.  Obviously, the node server accepting PROXY connections should not be exposed directly to the internet; only the proxy (whether ELB, HAProxy, or something else) should be able to connect to node.

API
---

### `proxy(Server[, options])`

Wraps something that inherits from the `net` module, exposing a `Server` and `createServer`.  Returns the same module patched to support the PROXY protocol.

Options:

- `strict` (default `true`): Incoming connections MUST use the PROXY protocol.  If the first five bytes received aren't `PROXY`, the connection will be dropped.  Disabling this option will allow connections that don't use the PROXY protocol (so long as the first bytes sent aren't `PROXY`).  Disabling this option poses a security risk; it should be enabled in production.

- `ignoreStrictExceptions` (default `false`): `strict` shutdowns your process with an error attached, meaning that if it isn't being caught on socket's `error` event, node will terminate process with an `uncaughtException`. This option tells `strict` methods to destroy sockets without providing the exception, so `node` ignores it. See [#11](https://github.com/findhit/proxywrap/issues/11) for more info.

- `overrideRemote` (default `true`): **findhit-proxywrap** overrides `socket.remoteAddress` and `socket.remotePort` for compability proposes. If you set this as `false`, your `socket.remoteAddress` and `socket.remotePort` will have the Address and Port of your **load-balancer** or whatever you are using behind your app. You can also access client's Address and Port by using `socket.clientAddress` and `socket.clientPort`.


Thanks
------

Huge thanks to [Josh Dague](https://github.com/daguej) for creating original [proxywrap](https://github.com/daguej/node-proxywrap).


CHANGELOG.md
CHANGELOG
=========

# Upcoming

# 0.3.10
- Added `.header` exposure on `error`
- Fixed TCP port on pr [#11](https://github.com/findhit/proxywrap/pull/11)

# 0.3.9
- Added an option to ignore strict generated exceptions while destroying socket,
  called `ignoreStrictExceptions`. Defaults to `false`. Reason on [#11](https://github.com/findhit/proxywrap/issues/11)

# 0.3.7
- Fixed destructed problem when on non-strict #7
- Updated dependent modules

# 0.3.6
- Fixed npm problem

# 0.3.5
- Implemented a better IPv6 detection approach
- Moved bluebird to dev dependencies #8
- Updated findhit-util lib to minor 2 updates

# 0.3.4
- Added RegExp Protocol detection

# 0.2.0 (2013-10-10)
- Added `options` parameter, ability to disable strict protocol checks.  Thanks to kylegetson.

# 0.1.2 (2013-08-01)
- Removed a `console.log` call that was accidentally left in.

# 0.1.1 (2013-07-31)
- Initial release.


test/tests.util.js
var ProxyWrap = require( '../index' ),
	Promise = require( 'bluebird' ),
	Util = require( 'findhit-util' ),

	protocols = {
		net: require( 'net' ),
		http: require( 'http' ),
		// https: require( 'https' ),
		// spdy: require( 'spdy' ).server
	}

	sinon = require( 'sinon' ),
	chai = require( 'chai' ),
	expect = chai.expect;

module.defaults = {
	fakeConnect: {
		protocol: 'TCP4',
		clientAddress: '10.10.10.1',
		clientPort: 12456,
		proxyAddress: '10.10.10.254',
		proxyPort: 80,

		autoCloseSocket: true,
		testAttributes: true,

		header: undefined,
		headerJoinCRLF: true,
	},
}

module.exports = {

	createServer: function ( p, options ) {
		var pc = protocols[ p ],
			proxy = ProxyWrap.proxy( pc, options ),

			server = proxy.createServer(),
			port = Math.floor( ( Math.random() * 5000 ) + 15000 ), // To be sure that the port is not beeing used on test side
			host = '127.0.0.1';

		server._protocol = p;
		server._protocolConstructor = pc;
		server.host = host;
		server.port = port;

		// Start server on localhost/random-port
		server.listen( port, host );

		// Returns server
		return server;
	},

	fakeConnect: function ( server, options ) {
		var header, body,
			p = server._protocol,
			pc = server._protocolConstructor;

		// Prepare options
		options = Util.extend( {}, module.defaults.fakeConnect, Util.is.Object( options ) && options || {} );

		// Build header
		header =
		(
			options.header ||
			[
				'PROXY',
				options.protocol,
				options.clientAddress,
				options.proxyAddress,
				options.clientPort,
				options.proxyPort,
			].join(' ')
		) +
		( options.headerJoinCRLF && "\r\n" || "" );

		body = [
			"GET /something/cool HTTP/1.1",
			"Host: www.findhit.com"
		].join("\n");

		return new Promise( function ( fulfill, reject ) {
			var client = new protocols.net.Socket(),

				host = server.host,
				port = server.port;

			var value = [ undefined, client ];

			server.once( 'connection', function ( socket ) {
				socket.on( 'error', function ( err ) {
					reject( err );
				});
			});

			server.once( 'proxiedConnection', function ( socket ) {
				value[0] = socket;

				socket.on( 'error', function ( err ) {
					reject( err );
				});

				if( options.testAttributes && ! options.header ) {
					try{
						expect( socket.clientAddress ).to.be.equal( options.clientAddress );
						expect( socket.proxyAddress ).to.be.equal( options.proxyAddress );
						expect( socket.clientPort ).to.be.equal( options.clientPort );
						expect( socket.proxyPort ).to.be.equal( options.proxyPort );

						if ( server.constructor.options.overrideRemote ) {
							expect( socket.remoteAddress ).to.be.equal( options.clientAddress );
							expect( socket.remotePort ).to.be.equal( options.clientPort );
						}

					} catch ( err ) {
						reject( err );
					}
				}

				if( options.autoCloseSocket ) {
					socket.end();
				} else {
					fulfill( value );
				}

			});

			client.once( 'connect', function () {
				// Send header and body
				client.write( header + body );
			});

			client.connect( port, host );

			if( options.autoCloseSocket ) {
				client.once( 'end', function () {
					fulfill( value );
				});
			};

		});
	}

};


test/proxy.protocol.v1.test.js
var Util = require( 'findhit-util' ),
	tUtil = require( './tests.util' ),
	Promise = require( 'bluebird' ),

	net = require( 'net' ),
	sinon = require( 'sinon' ),
	chai = require( 'chai' ),
	expect = chai.expect;

describe( "PROXY Protocol v1", function () {

	describe( 'net', function () {
		var server = tUtil.createServer( 'net', { strict: true });

		v1tests( 'net', server );
	});

	describe( 'http', function () {
		var server = tUtil.createServer( 'http', { strict: true });

		v1tests( 'http', server );
	});

	/*
	describe( 'https', function () {
		var server = tUtil.createServer( 'https', { strict: true });

		v1tests( server );
	});

	describe( 'spdy', function () {
		var server = tUtil.createServer( 'spdy', { strict: true });

		v1tests( server );
	});
	*/

});


function v1tests ( proto, server ) {

	it( "Check socket is stablished correctly", function () {
		return tUtil.fakeConnect( server );
	});

	it( "Check with another socket parameters", function () {
		return tUtil.fakeConnect(
			server,
			{
				clientAddress: '192.168.0.1',
				clientPort: 3350,
				proxyAddress: '192.168.0.254',
				proxyPort: 443,
			}
		);
	});

	it( "Check with another socket parameters as a string", function () {
		return tUtil.fakeConnect(
			server,
			{
				header: 'PROXY TCP4 192.168.0.1 192.168.0.254 3350 443',
			}
		);
	});

	describe( "Should detect a malformed PROXY headers", function () {

		it( "Header without IP's", function () {
			return tUtil.fakeConnect( server, {
				header: 'PROXY HACK ATTEMPT',
			})
				.then(function () {
					throw new Error("It shouldn't get fulfilled");
				}, function ( err ) {
					expect( err.message ).to.be.equal( "PROXY protocol malformed header" );
				});

		});

		if ( proto === 'net' ) {

			it( "non-proxy connection when in non-strict mode should not be destroyed #7", function () {
				return tUtil.fakeConnect( tUtil.createServer( proto, { strict: false }), {
					header: 'TELNET BABY',
				});
			});

		}

		it( "Restore emitted events after socket.destroy #5", function () {
			return tUtil.fakeConnect( server, {
				header: 'PRO',
				autoCloseSocket: false,
				testAttributes: false,
			})
				.then(function () {
					throw new Error("It shouldn't get fulfilled");
				}, function ( err ) {
					expect( err.message ).to.be.equal( "non-PROXY protocol connection" );
				});

		});

		it( "should drop connection gracefully when non-proxy connection is gathered when `ignoreStrictExceptions` is active. #11", function ( cb ) {
			var server = tUtil.createServer( proto, { strict: true, ignoreStrictExceptions: true });
			var client = new net.Socket();

			client.on( 'end', cb );

			client.once( 'connect', function () {
				// Send header and body
				client.write( 'GET / HTTP/1.0\n\n' );
			});

			client.connect( server.port, server.host );

		});

	});
}


lib/proxy-protocol.regexp.js
/*
[source](http://www.haproxy.org/download/1.5/doc/proxy-protocol.txt)

2.1

This is the format specified in version 1 of the protocol. It consists in one
line of ASCII text matching exactly the following block, sent immediately and
at once upon the connection establishment and prepended before any data flowing
from the sender to the receiver :

- a string identifying the protocol : "PROXY" ( \x50 \x52 \x4F \x58 \x59 )
Seeing this string indicates that this is version 1 of the protocol.

- exactly one space : " " ( \x20 )

- a string indicating the proxied INET protocol and family. As of version 1,
only "TCP4" ( \x54 \x43 \x50 \x34 ) for TCP over IPv4, and "TCP6"
( \x54 \x43 \x50 \x36 ) for TCP over IPv6 are allowed. Other, unsupported,
or unknown protocols must be reported with the name "UNKNOWN" ( \x55 \x4E
\x4B \x4E \x4F \x57 \x4E ). For "UNKNOWN", the rest of the line before the
CRLF may be omitted by the sender, and the receiver must ignore anything
presented before the CRLF is found. Note that an earlier version of this
specification suggested to use this when sending health checks, but this
causes issues with servers that reject the "UNKNOWN" keyword. Thus is it
now recommended not to send "UNKNOWN" when the connection is expected to
be accepted, but only when it is not possible to correctly fill the PROXY
line.

- exactly one space : " " ( \x20 )

- the layer 3 source address in its canonical format. IPv4 addresses must be
indicated as a series of exactly 4 integers in the range [0..255] inclusive
written in decimal representation separated by exactly one dot between each
other. Heading zeroes are not permitted in front of numbers in order to
avoid any possible confusion with octal numbers. IPv6 addresses must be
indicated as series of 4 hexadecimal digits (upper or lower case) delimited
by colons between each other, with the acceptance of one double colon
sequence to replace the largest acceptable range of consecutive zeroes. The
total number of decoded bits must exactly be 128. The advertised protocol
family dictates what format to use.

- exactly one space : " " ( \x20 )

- the layer 3 destination address in its canonical format. It is the same
format as the layer 3 source address and matches the same family.

- exactly one space : " " ( \x20 )

- the TCP source port represented as a decimal integer in the range
[0..65535] inclusive. Heading zeroes are not permitted in front of numbers
in order to avoid any possible confusion with octal numbers.

- exactly one space : " " ( \x20 )

- the TCP destination port represented as a decimal integer in the range
[0..65535] inclusive. Heading zeroes are not permitted in front of numbers
in order to avoid any possible confusion with octal numbers.

- the CRLF sequence ( \x0D \x0A )


The maximum line lengths the receiver must support including the CRLF are :
- TCP/IPv4 :
"PROXY TCP4 255.255.255.255 255.255.255.255 65535 65535\r\n"
=> 5 + 1 + 4 + 1 + 15 + 1 + 15 + 1 + 5 + 1 + 5 + 2 = 56 chars

- TCP/IPv6 :
"PROXY TCP6 ffff:f...f:ffff ffff:f...f:ffff 65535 65535\r\n"
=> 5 + 1 + 4 + 1 + 39 + 1 + 39 + 1 + 5 + 1 + 5 + 2 = 104 chars

- unknown connection (short form) :
"PROXY UNKNOWN\r\n"
=> 5 + 1 + 7 + 2 = 15 chars

- worst case (optional fields set to 0xff) :
"PROXY UNKNOWN ffff:f...f:ffff ffff:f...f:ffff 65535 65535\r\n"
=> 5 + 1 + 7 + 1 + 39 + 1 + 39 + 1 + 5 + 1 + 5 + 2 = 107 chars

*/

var IPv4 = '(((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?).){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?))',
    IPv6 = '([0-9a-fA-F]{0,4}:){2,7}(:|[0-9a-fA-F]{1,4})',
    Port = '([1-9][0-9]{0,3}|[1-5][0-9]{4}|6[0-4][0-9]{3}|65[0-2][0-9]{2}|653[0-2][0-9]|6533[0-5])'; // 1..65335


var ProxyProtocolRegexp = new RegExp(
    '^PROXY (' +
        'UNKNOWN' + '.{0,94}' + '|' + /* 107 - 13 */
        [ 'TCP4', IPv4, IPv4, Port, Port ].join(' ') +'|'+
        [ 'TCP6', IPv6, IPv6, Port, Port ].join(' ') +
    ')$'
);

// Export it
module.exports = ProxyProtocolRegexp;


lib/proxywrap.js
/*
 * node-proxywrap
 *
 * Copyright (c) 2013, Josh Dague
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

var util = require( 'util' ),
	Util = require( 'findhit-util' ),
	ProxyProtocolRegexp = require( './proxy-protocol.regexp.js' );

//var legacy = !require('stream').Duplex;  // TODO: Support <= 0.8 streams interface

exports.defaults = {
	strict: true,
	ignoreStrictExceptions: false,
	overrideRemote: true,
};

// Wraps the given module (ie, http, https, net, tls, etc) interface so that
// `socket.remoteAddress` and `remotePort` work correctly when used with the
// PROXY protocol (http://haproxy.1wt.eu/download/1.5/doc/proxy-protocol.txt)
// strict option drops requests without proxy headers, enabled by default to match previous behavior, disable to allow both proxied and non-proxied requests
exports.proxy = function(iface, options) {
	var exports = {};

	options = Util.extend( {}, module.exports.defaults, Util.is.Object( options ) && options || {} );

	// copy iface's exports to myself
	for (var k in iface) exports[k] = iface[k];


	function ProxiedServer(options, requestListener) {
		if (!(this instanceof ProxiedServer)) return new ProxiedServer(options, requestListener);

		if (typeof options == 'function') {
			requestListener = options;
			options = null;
		}

		// call original constructor with correct argument order
		if (options) iface.Server.call(this, options, requestListener);
		else iface.Server.call(this, requestListener);

		// remove the connection listener attached by iface.Server and replace it with our own.
		var cl = this.listeners('connection');
		this.removeAllListeners('connection');
		this.addListener('connection', connectionListener);

		// add the old connection listeners to a custom event, which we'll fire after processing the PROXY header
		for (var i = 0; i < cl.length; i++) {
			this.addListener('proxiedConnection', cl[i]);
		}

	}
	util.inherits(ProxiedServer, iface.Server);

	exports.createServer = function(opts, requestListener) {
		return new ProxiedServer(opts, requestListener);
	}

	exports.Server = ProxiedServer;

	exports.options =
	ProxiedServer.options =
		options;

	function connectionListener(socket) {
		var self = this, realEmit = socket.emit, history = [], protocolError = false;

		// TODO: Support <= 0.8 streams interface
		//function ondata() {}
		//if (legacy) socket.once('data', ondata);

		// override the socket's event emitter so we can process data (and discard the PROXY protocol header) before the underlying Server gets it
		socket.emit = function(event, data) {
			history.push(Array.prototype.slice.call(arguments));

			if (event == 'readable') {
				onReadable();
			}
		}

		function restore() {
			if( socket.emit === realEmit ) return;

			//if (legacy) socket.removeListener('data', ondata);
			// restore normal socket functionality, and fire any events that were emitted while we had control of emit()
			socket.emit = realEmit;
			for (var i = 0; i < history.length; i++) {
				realEmit.apply(socket, history[i]);
				if (history[i][0] == 'end' && socket.onend) socket.onend();
			}
			history = null;
		}

		function destroy( error ) {
			error = error || undefined;

			if ( ! ( error instanceof Error ) ) {
				error = new Error( error );
			}

			// Set header on error
			error.header = header;

			protocolError = true;

			socket.destroy( ! options.ignoreStrictExceptions ? error : undefined );

			restore();

		}

		socket.on('readable', onReadable);

		var header = '', buf = new Buffer(0);

		function onReadable() {
			var chunk;
			while (null != (chunk = socket.read())) {
				buf = Buffer.concat([buf, chunk]);
				header += chunk.toString('ascii');

				// if the first 5 bytes aren't PROXY, something's not right.
				if (header.length >= 5 && header.substr(0, 5) != 'PROXY') {
					protocolError = true;
					if (options.strict) {
						return destroy( "non-PROXY protocol connection" );
					}
				}

				var crlf = header.indexOf('\r');
				if (crlf > 0 || protocolError) {
					socket.removeListener('readable', onReadable);
					header = header.substr(0, crlf);

					// Check if header is valid
					if (options.strict) {
						if( ! ProxyProtocolRegexp.test( header ) ) {
							return destroy( "PROXY protocol malformed header" );
						}
					}

					var hlen = header.length;
					header = header.split(' ');

					if ( ! protocolError ) {

						if( options.overrideRemote ) {

							Object.defineProperty(socket, 'remoteAddress', {
								enumerable: false,
								configurable: true,
								get: function() {
									return header[2];
								}
							});
							Object.defineProperty(socket, 'remotePort', {
								enumerable: false,
								configurable: true,
								get: function() {
									return parseInt(header[4], 10);
								}
							});

						}

						// Source properties of TCP connection
						Object.defineProperty(socket, 'clientAddress', {
							enumerable: false,
							configurable: true,
							get: function() {
								return header[2];
							}
						});
						Object.defineProperty(socket, 'clientPort', {
							enumerable: false,
							configurable: true,
							get: function() {
								return parseInt(header[4], 10);
							}
						});

						// Destination properties of TCP connection
						Object.defineProperty(socket, 'proxyAddress', {
							enumerable: false,
							configurable: true,
							get: function() {
								return header[3];
							}
						});
						Object.defineProperty(socket, 'proxyPort', {
							enumerable: false,
							configurable: true,
							get: function() {
								return parseInt(header[5], 10);
							}
						});
					}

					// unshifting will fire the readable event
					socket.emit = realEmit;
					socket.unshift(buf.slice(protocolError ? 0 : crlf+2));

					self.emit('proxiedConnection', socket);

					restore();

					if ( socket.ondata ) {
						var data = socket.read();

						if ( data ) {
							socket.ondata(data, 0, data.length);
						}
					}

					break;

				} else if ( header.length > 107 ) {
					return destroy('PROXY header too long');
				}
			}
		}

	}

	return exports;
}
